src/main/java/evaluation/optimisation/NTBEA.java

```diff
@@ -136,92 +135,55 @@ public void writeAgentJSON(int[] settings, String fileName) {
      * @return
      */
     public Pair<Object, int[]> run() {
+        ExecutorService executor = params.nThreads > 1 ? Executors.newFixedThreadPool(Math.min(params.nThreads, params.repeats)) : null;
+        // if we're multithreading, we don't want to have different threads interfering with eachother
+        List<NTBEA> clones = new ArrayList<>();
+        if (executor == null) {
+            // no multithreading, so list of clones consists only of the current object
+            clones.add(this);
+        }
 
+        // Only this loop is parallelized, since the rest is just analysis, or initializing a tournament (which is
+        // on its own already parallelized)
         for (currentIteration = 0; currentIteration < params.repeats; currentIteration++) {
-            runIteration();
-            writeAgentJSON(winnerSettings.get(winnerSettings.size() - 1),
-                    params.destDir + File.separator + "Recommended_" + currentIteration + ".json");
+            NTBEA clone = executor == null ? this : this.copy();
+            if (executor != null) {
+                clone.currentIteration = currentIteration; // for correct reporting
+                // run in parallel if allowed
+                executor.submit(clone::runIteration);
+                clones.add(clone);
+            } else {
+                clone.runIteration();
+            }
         }
 
-        // After all runs are complete, if tournamentGames are specified, then we allow all the
-        // winners from each iteration to play in a tournament and pick the winner of this tournament
-        if (params.tournamentGames > 0 && winnersPerRun.get(0) instanceof AbstractPlayer) {
```

Comment:
This block and the next block of removed lines are all moved into the activateTournament() method, for legibility. This was done in a separate commit, to make it easier to follow all of the other canges made to this part of the code by looking at the other commits.